src/System.IO.Pipelines/PipelineConnectionStream.cs

```diff
@@ -0,0 +1,199 @@
+﻿using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO.Pipelines
+{
+    public class PipelineConnectionStream : Stream
+    {
+        private readonly static Task<int> _initialCachedTask = Task.FromResult(0);
+        private Task<int> _cachedTask = _initialCachedTask;
+
+        private readonly IPipelineConnection _connection;
+
+        public PipelineConnectionStream(IPipelineConnection connection)
+        {
+            _connection = connection;
+        }
+
+        public override bool CanRead => true;
+
+        public override bool CanSeek => false;
+
+        public override bool CanWrite => true;
+
+        public override long Length
+        {
+            get
+            {
+                ThrowHelper.ThrowNotSupportedException();
+                return 0;
+            }
+        }
+
+        public override long Position
+        {
+            get
+            {
+                ThrowHelper.ThrowNotSupportedException();
+                return 0;
+            }
+            set
+            {
+                ThrowHelper.ThrowNotSupportedException();
+            }
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            ThrowHelper.ThrowNotSupportedException();
+            return 0;
+        }
+
+        public override void SetLength(long value)
+        {
+            ThrowHelper.ThrowNotSupportedException();
+        }
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            // ValueTask uses .GetAwaiter().GetResult() if necessary
+            // https://github.com/dotnet/corefx/blob/f9da3b4af08214764a51b2331f3595ffaf162abe/src/System.Threading.Tasks.Extensions/src/System/Threading/Tasks/ValueTask.cs#L156
+            return ReadAsync(new ArraySegment<byte>(buffer, offset, count)).Result;
+        }
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            var task = ReadAsync(new ArraySegment<byte>(buffer, offset, count));
+
+            if (task.IsCompletedSuccessfully)
+            {
+                if (_cachedTask.Result != task.Result)
+                {
+                    // Needs .AsTask to match Stream's Async method return types
+                    _cachedTask = task.AsTask();
+                }
+            }
+            else
+            {
+                // Needs .AsTask to match Stream's Async method return types
+                _cachedTask = task.AsTask();
```

Comment:
>_cachedTask = task.AsTask(); [](start = 16, length = 28)

I'm not sure this is a good thing to do.  First, the operation could fail, in which case you've cached a task that has an exception, and the next time you check _cachedTask.Result, it'll throw.  Second, you could end up in a situation where a call to ReadAsync blocks waiting on the _cachedTask.  I think you only want to cache it in the task.IsCompletedSuccessfully case.